### ISSUE TITLE
uses may prove a can of worms

### ISSUE THREAD

**Issue by gardea** [raml-org/raml-spec#204]
I'm probably still absorbing the implications of uses. Still, I can't but wonder what real problem is it solving here; for sure, I see the same problems we've had with XML Schema at least (I loosely associate uses with include/import mechanism).

For e.g., let's look at the example provided in the spec under Applying Libraries; what if, in the files-resource.raml I also add: 

    uses:
      types: !include libraries/file-type.raml
      files: !include libraries/files.raml

I could then have, as per current example:
    type: files.file-type.File

or, per modified one:
    type: types.File

> In this way the library name defines a namespace for the library's objects within the context in which the library was applied.

The above could be thought very similar to how XSD created the chameleon pattern, where the same type assumed different qualified name by having a namespace-less schema included in a schema with a namespace.

Potential problems here?
 - How would code generation deal with these "namespaces"? 
 - What is the resultant symbol space here? Should they be treated as different types? What would the most common expectation be here? 
 - Should we check for "type already defined"? etc.

**Comment by sichvoge** [comment]
Hi Petru,

to your 'potential problems' list and also to try keep that discussion alive:

1. code generation shouldn't be a problem for most languages that have 'namespaces' as a concept as well. In JS where you not really have that you could define functions and variables (please note, i am not a js expert) for example. I have seen couple of posts how to do that googling the problem.

2. really depends on how you resolve them, if you for the current approach using prefixes, that could help to resolve the symbol space by treating them as two different types

```yaml
uses:
  ext1: !include lib1.raml
  ext2: !include lib2.raml

/resource1:
  get:
    body: 
      application/json:
        type: ext1.People
/resource2:
  get:
    body: 
      application/json:
        type: ext2.People    
```

If you use another approach described in #207 where you don't have a prefix, you need to throw an error with identical types in both libraries

```yaml
uses:
  ~: !include lib1.raml#People
  ~: !include lib2.raml#People

/resource1:
  get:
    body: 
      application/json:
        type: People
/resource2:
  get:
    body: 
      application/json:
        type: People
```

or another use case where you just import everything

```yaml
uses:
  ~: !include lib1.raml
  ~: !include lib2.raml

/resource1:
  get:
    body: 
      application/json:
        type: People
/resource2:
  get:
    body: 
      application/json:
        type: People
```

you would need to throw an error on the `type`s to indicate that you need to define them explicitly using prefixes again.

+3. see answer to 2.

Just an idea and hope people will discuss here as well.

**Comment by sichvoge** [comment]
related to #226

**Comment by usarid** [comment]
@gardea I'm hoping that #226 does indeed address the various issues in this issue -- let us know and we can look to close this one.

**Comment by sichvoge** [comment]
Addressed by #226 and already put into RC2
